contracts/PooledStaking.sol

```diff
@@ -0,0 +1,512 @@
+/*
+    Copyright (C) 2020 NexusMutual.io
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/
+*/
+
+pragma solidity ^0.5.16;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+import "./base/MasterAware.sol";
+import "./base/TokenAware.sol";
+import "./libraries/Vault.sol";
+
+contract PooledStaking is MasterAware, TokenAware {
+  using SafeMath for uint;
+
+  enum ParamType {
+    MIN_STAKE,
+    MAX_LEVERAGE,
+    MIN_ALLOWED_DEALLOCATION,
+    DEALLOCATE_LOCK_TIME
+  }
+
+  struct Staker {
+    uint staked; // total amount of staked nxm
+    uint reward; // total amount that is ready to be claimed
+    address[] contracts; // list of contracts the staker has staked on
+
+    mapping(address => uint) allocations;
+
+    // amount to be subtracted after all deallocations will be processed
+    mapping(address => uint) pendingDeallocations;
+  }
+
+  struct Contract {
+    uint staked; // amount of nxm staked for this contract
+    // TODO: find a way to remove zero-amount stakers. Linked list?
+    address[] stakers; // used for iteration
+  }
+
+  struct Burn {
+    uint amount;
+    uint burnedAt;
+    address contractAddress;
+    uint next; // id of the next deallocation request in the linked list
```

Comment:
to clarify the reasons for using he linked list here are:

* the list is sorted by deallocation time
* O(1) insertion into said list using index hint